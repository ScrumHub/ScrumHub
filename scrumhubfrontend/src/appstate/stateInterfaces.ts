import { tableKeys } from "../components/utility/commonInterfaces";

export type Error = {
  hasError: boolean;
  errorCode: number;
  erorMessage: string;
};
export interface IFilters {
  [name: string]: any|any[];
}
export interface IFiltersAndToken {
  filters: IFilters;
  token: string;
}

export interface IError {
  data: any;
  Message: string;
  metadata: any;
  successful: boolean;
  title?:string;
}

export interface ITask {
  assigness: IPerson[]|any[];
  id: number;
  name: string;
  finished: boolean;
  pbiId: number;
  isAssignedToPBI: boolean;
  link: string;
  status:string;
}

export interface ITaskList extends IFetchedList {
  list: ITask[];
}

export interface IBacklogItem {
  id: number;
  name: string;
  finished: boolean;
  expectedTimeInHours: number;
  estimated: boolean;
  sprintNumber: number | null;
  isInSprint: boolean;
  timeSpentInHours: number;
  priority: number;
  acceptanceCriteria: string[];
  tasks: ITask[]
}

export interface IAddBI {
  name: string;
  priority: number;
  acceptanceCriteria: string[];
}
export interface IAssignBI {
  name: string;
  id: number;
  isInSprint:boolean;
}
export interface IBIFilter {
  pageNumber: number;
  pageSize: number;
  nameFilter?: string;
  finished?: boolean;
  estimated?: boolean;
  inSprint?: boolean;
}

export interface IFetchedList {
  pageNumber: number;
  pageCount: number;
  pageSize: number;
  realSize: number;
}

export interface IBacklogItemList extends IFetchedList{
  list: IBacklogItem[];
}

export interface ISprint {
  sprintNumber: number;
  title:string;
  goal: string;
  backlogItems: IBacklogItem[];
  finishDate: string|Date;
  isCurrent: boolean;
  status: string;
  isCompleted: boolean;
}

export interface ISprintList extends IFetchedList{
  list: ISprint[];
}

export interface IRepository {
  name: string;
  description: string;
  dateOfLastActivity: Date | string | null;
  typeOfLastActivity: string;
  gitHubId: number;
  hasAdminRights: boolean;
  alreadyInScrumHub: boolean;
  backlogItems: IBacklogItem[] | any;
  sprints: ISprint[] | any;
}

export interface IPerson {
  avatarLink:string;
  gitHubId: number;
  isCurrentUser:boolean;
  login: string;
  name: string|null;
}

export interface IPeopleList extends IFetchedList{
  list: IPerson[];
}

export type BodyRowTypes = ISprint | IBacklogItem | ITask;
export interface IRepositoryList extends IFetchedList{
  list: IRepository[];
}
export interface ILoginState {
  isLoggedIn: boolean,
  token: string,
  client_id: string,
  redirect_uri: string,
  client_secret: string,
  proxy_url:string
}
export interface IState {
  loginState:ILoginState;
  loading: boolean;
  error: Error;
  pbiPage: IBacklogItemList;
  repositories: IRepositoryList | any;
  reposLastPage: boolean;
  reposRequireRefresh: boolean;
  productLastPage: boolean;
  productRequireRefresh: boolean;
  sprintLastPage: boolean;
  sprintRequireRefresh: boolean;
  sprintPage: ISprintList;
  taskLastPage: boolean;
  taskRequireRefresh: boolean;
  taskPage: ITaskList;
  tasks: ITask[];
  openSprint: ISprint | null;
  repoId: number;
  namedPBI: IAssignBI[];
  people: IPeopleList;
  activeSprintNumber:number;
  currentUser: IPerson;
  keys:tableKeys;
  loadingKeys:tableKeys;
  changedRepo:string;
};