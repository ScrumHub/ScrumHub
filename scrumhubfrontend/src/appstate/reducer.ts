import * as Actions from "./actions";
import { createReducer, PayloadAction } from "@reduxjs/toolkit";
import { RequestResponse } from "./response";
import { IAssignBI, ILoginState, IPeopleList, IPerson, IBacklogItem, IBacklogItemList, IRepository, IRepositoryList, ISprint, ISprintList, ITask, ITaskList, IState } from "./stateInterfaces";
import { initError, loggedOutLoginState, unassignedPBI } from "./stateInitValues";
import { isArrayValid, isNameFilterValid } from "../components/utility/commonFunctions";
import { getError, updateStateTasks, updateStateKeys, addStateTask, addStateUnassignedTaskToPBI, updateStatePBI, updateTasksSWR, fetchStateRepos, updateStateOneSprint, addStateRepo, addStatePBI, addStateSprint, updateOnDragStateTasks } from "./stateUtilities";
import { isNull } from "lodash";
import config from "../configuration/config";
var _ = require('lodash');

export const reducerFunction = (initState: IState) => {
  return (createReducer(initState, {
    [Actions.clearError.type]: (state: IState, payload: any) => {
      let newState = _.cloneDeep(state as IState);
      return ({ ...newState, error: initError, changedRepo:isNameFilterValid(payload.payload)?payload.payload:""});
    },
    [Actions.clearReposList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.repositories = [];
      newState.reposRequireRefresh = true;
      newState.reposLastPage = false;
      return newState;
    },
    [Actions.clearState.type]: (state: IState) => {
      return (initState);
    },
    [Actions.clearProject.type]: (state: IState) => {
      let newState = _.cloneDeep(state) as IState;
      return { ...initState, repositories: newState.repositories, loginState:newState.loginState, currentUser: newState.currentUser };
    },
    [Actions.clearPBIsList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.pbiPage.list = [];
      newState.productRequireRefresh = true;
      newState.reposLastPage = false;
      return newState;
    },
    [Actions.clearSprintList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.sprintPage.list = [];
      newState.sprintRequireRefresh = true;
      newState.sprintLastPage = false;
      return newState;
    },
    [Actions.clearSprint.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.openSprint = null;
      newState.sprintRequireRefresh = true;
      newState.sprintLastPage = false;
      return newState;
    },
    [Actions.updateSprintKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, keys: { ...newState.keys, sprintKeys: updateStateKeys(newState.keys.sprintKeys, temp) } };
    },
    [Actions.updatePBIKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, keys: { ...newState.keys, pbiKeys: updateStateKeys(newState.keys.pbiKeys, temp) } };
    },
    [Actions.updateSprintLoadingKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, loadingKeys: { ...newState.loadingKeys, sprintKeys: updateStateKeys(newState.loadingKeys.sprintKeys, temp) } };
    },
    [Actions.updatePBILoadingKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, loadingKeys: { ...newState.loadingKeys, pbiKeys: updateStateKeys(newState.loadingKeys.pbiKeys, temp) } };
    },
    [Actions.login.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const tempLogin = expanded.payload as ILoginState
      return ({ ...newState, loginState: { ...newState.loginState, isLoggedIn: tempLogin.isLoggedIn, token: tempLogin.token } });
    },
    [Actions.logout.type]: (state: IState) => { return ({ ...initState, loginState: loggedOutLoginState }); },
    /*REPOS*/
    [Actions.fetchReposThunk.pending.toString()]: (state: IState) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchReposThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IRepositoryList, number>>) => {
      let newState = _.cloneDeep(state);
      return fetchStateRepos(newState, payload.payload.response as IRepositoryList, _.get(payload, ["meta", "arg", "filters", "pageNumber"], config.defaultFilters.page),
       _.get(payload, ["meta", "arg", "filters", "pageSize"], config.defaultFilters.size),_.get(payload.payload.response, ["pageCount"], 1));
    },
    [Actions.fetchReposThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addRepositoryThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addRepositoryThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IRepository, number>>) => {
      let newState = _.cloneDeep(state);
      return (addStateRepo(newState, payload.payload.response as IRepository));
    },
    [Actions.addRepositoryThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*PBIS*/
    [Actions.finishPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.finishPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.finishPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return(addStatePBI(newState,payload.payload.response as IBacklogItem));
    },
    [Actions.addPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.deletePBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.deletePBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<any, any>>) => {
      let newState = _.cloneDeep(state);
      newState.loading = false;
      return newState;
    },
    [Actions.deletePBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchPBIsThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPBIsThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItemList, number>>) => {
      let newState = _.cloneDeep(state);
      const pbisList = payload.payload.response as IBacklogItemList;
      if (newState.pbiPage && isArrayValid(newState.pbiPage.list) && newState.pbiPage.list[0].id === 0) {
        newState.pbiPage = { ...pbisList, list: [newState.pbiPage.list[0]].concat(pbisList.list) };
      }
      else {
        newState.pbiPage = { ...pbisList, list: [unassignedPBI].concat(pbisList.list) };
      }
      newState.loading = false;
      return newState;
    },
    [Actions.fetchPBIsThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchPeopleThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPeopleThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IPeopleList, number>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, people: payload.payload.response as IPeopleList, error: initError, loading: false };
    },
    [Actions.fetchPeopleThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.getCurrentUserThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.getCurrentUserThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IPerson, number>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, currentUser: payload.payload.response as IPerson, error: initError, loading: false };
    },
    [Actions.getCurrentUserThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.getPBINamesThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.getPBINamesThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItemList, number>>) => {
      let newState = _.cloneDeep(state);
      newState.error = initError;
      newState.namedPBI = (payload.payload.response as IBacklogItemList).list as IAssignBI[];
      newState.loading = false;
      return newState;
    },
    [Actions.getPBINamesThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.estimatePBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.estimatePBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.estimatePBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.editPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.editPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.editPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*SPRINTS*/
    [Actions.fetchSprintsThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchSprintsThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprintList, number>>) => {
      let newState = _.cloneDeep(state);
      const sprints = payload.payload.response as ISprintList;
      newState.sprintPage = sprints;
      const index = sprints && isArrayValid(sprints.list) ? sprints.list.findIndex((sprint: ISprint) => sprint.isCurrent) : -1;
      if (index !== -1) { newState.activeSprintNumber = sprints.list[index].sprintNumber };
      newState.sprintRequireRefresh = false;
      newState.loading = false;
      newState.error = initError;
      return newState;
    },
    [Actions.fetchSprintsThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, number>>) => {
      let newState = _.cloneDeep(state);
      newState.openSprint = payload.payload.response as ISprint;
      if (!isNull(newState.openSprint) && newState.openSprint.isCurrent) { newState.activeSprintNumber = newState.openSprint.sprintNumber };
      newState.sprintRequireRefresh = false;
      newState.error = initError;
      newState.loading = false;
      return newState;
    },
    [Actions.fetchOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.updateOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.updateOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint | any, number>>) => {
      let newState = _.cloneDeep(state);
      return (updateStateOneSprint(newState, payload.payload.response as ISprint));
    },
    [Actions.updateOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.completeOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.completeOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, number>>) => {
      let newState = _.cloneDeep(state);
      return (updateStateOneSprint(newState, payload.payload.response as ISprint));
    },
    [Actions.completeOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, undefined>>) => {
      let newState = _.cloneDeep(state);
      return(addStateSprint(newState,payload.payload.response as ISprint));
    },
    [Actions.addSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*TASKS*/
    [Actions.fetchPBITasksThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPBITasksThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITaskList, number>>) => {
      let newState = _.cloneDeep(state);
      return updateTasksSWR(newState, payload.payload.response as ITaskList);
    },
    [Actions.updateTasks.type]: (state: IState, payload: PayloadAction<any>) => {
      let newState = _.cloneDeep(state);
      const item = payload.payload as IBacklogItem;
      return updateStatePBI(newState, item);
    },
    [Actions.fetchPBITasksThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addUnassignedTasksToPBI.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addUnassignedTasksToPBI.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITaskList, number>>) => {
      let newState = _.cloneDeep(state);
      return addStateUnassignedTaskToPBI(newState, payload.payload.response as ITaskList);
    },
    [Actions.addUnassignedTasksToPBI.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return addStateTask(newState, payload.payload.response as ITask);
    },
    [Actions.addTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.assignTaskToPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.assignTaskToPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateOnDragStateTasks(newState,payload.payload.response as ITask);
    },
    [Actions.assignTaskToPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.updatePersonInTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.updatePersonInTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStateTasks(newState, payload.payload.response as ITask);
    },
    [Actions.updatePersonInTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.startTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.startTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStateTasks(newState, payload.payload.response as ITask);
    },
    [Actions.startTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
  }));
}