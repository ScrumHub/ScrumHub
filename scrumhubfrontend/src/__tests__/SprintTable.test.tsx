/**
 * @jest-environment jsdom
 */
import { configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { isNull } from 'lodash';
import { unmountComponentAtNode } from 'react-dom';
import { shallow } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';
import { initFilteredInfo, initModalVals, initSortedInfo } from '../components/utility/commonInitValues';
import { initSprint } from '../appstate/stateInitValues';
import { dragCmpnts, pbiColumns, sprintColumns } from '../components/utility/TableUtilities';
import { TestDraggableBodyRow } from '../components/utility/BacklogHandlers';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { reducerFunction } from '../appstate/reducer';
import { updatedTestState } from '../appstate/stateTestValues';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { DndProvider } from 'react-dnd';
import TestRenderer from "react-test-renderer";
import { PBITableComponent } from '../components/tables/PBITable';
import config from '../configuration/config';
import { SprintTableComponent } from '../components/tables/SprintTable';
import { ISprint } from '../appstate/stateInterfaces';
const { act: actTester } = TestRenderer;
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn()
}));


configure({ adapter: new Adapter() });
describe('SprintTable component in container', () => {
  let tempContainer: any, setState: any, store: any, useStateMock: any;
  beforeEach(() => {
    tempContainer = document.createElement('div');
    document.body.appendChild(tempContainer);
    store = configureStore({
      reducer: reducerFunction({ ...updatedTestState, loading: true }),
    });
    jest.spyOn(console, 'warn').mockImplementation(() => { });
    jest.spyOn(console, 'error').mockImplementation(() => { });
    setState = jest.fn();
    useStateMock = (initState: any) => [initState, setState];
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  afterEach(() => {
    // cleanup on exiting
    if (!isNull(tempContainer)) {
      unmountComponentAtNode(tempContainer);
      tempContainer.remove();
      tempContainer = null;
    }
  });
  it('is the same as snapshot', () => {
    let navigate = jest.fn();
    const PBITableforSprint: React.FC<ISprint> = (item: ISprint) => {
      return (
        <PBITableComponent sortedInfo={initSortedInfo} filteredInfo={initFilteredInfo}
          sortSelected={function (items: any): void { }}
          itemSelected={function (items: number[]): void { }} TaskTableforPBI={[]}
          nameFilter={[]} peopleFilter={[]} item={initSprint}
          pbiColumns={pbiColumns([], initSortedInfo, initFilteredInfo, useStateMock,
            initModalVals, useStateMock, [] as number[], config.token, "ownerName")}
          nestedcomponents={dragCmpnts(TestDraggableBodyRow)} />)
    };

    const subject = shallow(<Provider store={store}>
      <DndProvider backend={HTML5Backend} key={"dnd_sprint"}>
        <SprintTableComponent
          loading={true}
          data={[initSprint]} components={dragCmpnts(TestDraggableBodyRow)}
          columns={sprintColumns("ownerName", navigate, initSortedInfo, initFilteredInfo,
            useStateMock, initModalVals, useStateMock)}
          PBITableforSprint={PBITableforSprint}
          sortedInfo={initSortedInfo} filteredInfo={initFilteredInfo}
          nameFilter={[]} peopleFilter={[]} />
      </DndProvider>
    </Provider>);
    expect(EnzymeToJson(subject)).toMatchSnapshot();
  });
});

