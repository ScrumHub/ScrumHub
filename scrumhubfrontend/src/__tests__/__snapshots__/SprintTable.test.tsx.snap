// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintTable component in container is the same as snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Memo(DndProvider)
    backend={[Function]}
    key="dnd_sprint"
  >
    <Memo()
      PBITableforSprint={[Function]}
      columns={
        Array [
          Object {
            "align": "left",
            "key": "sprintNumber",
            "render": [Function],
            "title": "Number",
            "width": "15%",
          },
          Object {
            "align": "center",
            "dataIndex": "title",
            "ellipsis": true,
            "key": "sprintTitle",
            "title": "Title",
            "width": "27%",
          },
          Object {
            "align": "center",
            "dataIndex": "finishDate",
            "key": "finishDate",
            "render": [Function],
            "title": "Deadline",
            "width": "15%",
          },
          Object {
            "align": "center",
            "key": "finishDate",
            "render": [Function],
            "title": "Story Points",
            "width": "15%",
          },
          Object {
            "align": "center",
            "filteredValue": Array [],
            "filters": Array [
              Object {
                "text": "Complete",
                "value": 1,
              },
              Object {
                "text": "Not Complete",
                "value": 0,
              },
            ],
            "key": "isCompleted",
            "onFilter": [Function],
            "render": [Function],
            "sortOrder": false,
            "sorter": [Function],
            "title": "completed",
            "width": "15%",
          },
          Object {
            "align": "right",
            "key": "action",
            "render": [Function],
            "title": "Action",
            "width": "8%",
          },
        ]
      }
      components={
        Object {
          "body": Object {
            "row": [Function],
          },
        }
      }
      data={
        Array [
          Object {
            "backlogItems": Array [
              Object {
                "acceptanceCriteria": Array [
                  "criteria",
                  "criteria2",
                ],
                "estimated": true,
                "expectedTimeInHours": 2,
                "finished": false,
                "id": 0,
                "isInSprint": false,
                "name": "Item",
                "priority": 0,
                "sprintNumber": 0,
                "tasks": Array [
                  Object {
                    "assigness": Array [],
                    "finished": false,
                    "id": 0,
                    "isAssignedToPBI": false,
                    "link": "",
                    "name": "",
                    "pbiId": 0,
                    "status": "",
                  },
                ],
                "timeSpentInHours": 0,
              },
            ],
            "finishDate": "",
            "goal": "",
            "isCompleted": false,
            "isCurrent": false,
            "sprintNumber": 1,
            "status": "NotFinished",
            "title": "",
          },
        ]
      }
      filteredInfo={
        Object {
          "complete": Array [],
          "pbiPriority": Array [],
        }
      }
      loading={true}
      nameFilter={Array []}
      peopleFilter={Array []}
      sortedInfo={
        Object {
          "columnKey": "",
          "order": "",
        }
      }
    />
  </Memo(DndProvider)>
</ContextProvider>
`;
